src/python/pants/backend/scala/resolve/lockfile_test.py

```diff
@@ -0,0 +1,90 @@
+# Copyright 2022 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from __future__ import annotations
+
+import pytest
+
+from pants.backend.scala.dependency_inference import rules as scala_dep_inf_rules
+from pants.backend.scala.resolve.lockfile import rules as scala_lockfile_rules
+from pants.core.goals.generate_lockfiles import GenerateLockfileResult
+from pants.core.util_rules import archive, source_files
+from pants.core.util_rules.external_tool import rules as external_tool_rules
+from pants.engine.fs import DigestContents, FileDigest
+from pants.engine.internals import graph
+from pants.jvm import jdk_rules
+from pants.jvm.goals import lockfile
+from pants.jvm.goals.lockfile import RequestedJVMserResolveNames
+from pants.jvm.resolve.common import ArtifactRequirements, Coordinate, Coordinates
+from pants.jvm.resolve.coursier_fetch import CoursierLockfileEntry, CoursierResolvedLockfile
+from pants.jvm.resolve.coursier_fetch import rules as coursier_fetch_rules
+from pants.jvm.resolve.coursier_setup import rules as coursier_setup_rules
+from pants.jvm.resolve.jvm_tool import rules as coursier_jvm_tool_rules
+from pants.jvm.resolve.lockfile_metadata import JVMLockfileMetadata
+from pants.jvm.target_types import JvmArtifactTarget
+from pants.jvm.testutil import maybe_skip_jdk_test
+from pants.jvm.util_rules import rules as util_rules
+from pants.testutil.rule_runner import QueryRule, RuleRunner
+
+
+@pytest.fixture
+def rule_runner() -> RuleRunner:
+    rule_runner = RuleRunner(
+        rules=[
+            *scala_lockfile_rules(),
+            *scala_dep_inf_rules.rules(),
+            *jdk_rules.rules(),
+            *coursier_fetch_rules(),
+            *coursier_jvm_tool_rules(),
+            *lockfile.rules(),
+            *coursier_setup_rules(),
+            *external_tool_rules(),
+            *source_files.rules(),
+            *util_rules(),
+            *archive.rules(),
+            *graph.rules(),
+            # QueryRule(UserGenerateLockfiles, [RequestedJVMserResolveNames]),
+            QueryRule(GenerateLockfileResult, [RequestedJVMserResolveNames]),
```

Comment:
@stuhood: This is the test with the rule graph errors, not https://github.com/pantsbuild/pants/blob/11fd5ee442acfbce733088502c8288d04bedd1bf/src/python/pants/jvm/goals/lockfile_test.py#L42-L43.

Reply:
Yes. And these `QueryRules` are different from the queries used there, which is the root cause of the issue. AFAICT, a `GenerateLockfileResult` cannot be computed from a `RequestedJVMserResolveNames` (only a `UserGenerateLockfiles` can be).

Reply:
Ah doh. Too many types.